doc/book/cookbook/setting-locale-depending-routing-parameter.md

```diff
@@ -0,0 +1,303 @@
+# How can I setup the locale depending on a routing parameter?
+
+It is a common task to have an localized web application where the setting of
+the locale (and therefor the language) depends on the routing. There are other 
+ways of setting a language like using the session or a specialized sub-domain.
+In this recipe we will concentrate on using the a routing parameter.
+
+## Setting up the route ##
+
+If you want to set the locale depending on an routing parameter, you first have
+to add a language parameter to your routes. In this example we use the `lang` 
```

Comment:
@RalfEggert — there's another way to do this that doesn't require touching each and every route: pre_routing pipeline middleware.

As an example, you would define such middleware like this:

```php
$uri = $request->getUri();
$path = $uri->getPath();
if (! preg_match('#^/(?P<lang>[a-z]{2})/#', $path, $matches) {
    Locale::setDefault('de_DE');
    return $next($request, $response);
}

$lang = $matches['lang'];
Locale::setDefault($lang);

$path = substr($path, 3);
$request = $request
    ->withUri($uri->withPath($path))
    ->withAttribute('lang', $lang);
return $next($request, $response);
```

This approach has a couple of benefits:

- No need for a route result observer.
- No need to prefix each and every route with `/:lang`.

The primary problem with it will be URI generation, as the router will be generating URIs without the language prefix, which will require a custom URI helper. However, this could even be something as simple as:

```php
namespace My\Application;

use Locale;
use Zend\Expressive\ZendView\UrlHelper;

class LocalizedUrlHelper extends UrlHelper
{
    public function __invoke($route = null, $params = [])
    {
        return sprintf('/%s%s', Locale::getDefault(), parent::__invoke($route, $params));
    }
}
```

I'm personally of the mind that we should demonstrate both approaches in the cookbook. Thoughts?